internal/replacer/replacer.go

```diff
@@ -190,51 +200,56 @@ func (r *Redactor) Flush() error {
 }
 
 // flush writes out the buffer up to an index. limit is an upper limit.
-func (r *Redactor) flushUpTo(limit int) error {
+func (r *Replacer) flushUpTo(limit int) error {
 	if limit == 0 || len(r.buf) == 0 {
 		return nil
 	}
 
 	bufidx := 0 // where we are up to in the buffer
 	done := -1  // the index of the last match processed
 
-	// Stop when we're out of redactions, or the next one is after limit.
+	// Stop when we're out of completed matches, or the next one is after limit.
 
 	for ri, match := range r.completedMatches {
 		if match.from >= limit {
 			// This range is after the cutoff point.
 			break
 		}
-		done = ri
 
-		switch {
-		case bufidx < match.from:
-			// A non-redacted range (followed by a redacted range).
+		if match.to > limit {
+			// This range overlaps the cutoff point. Adjust the limit in order
+			// to be able to give the complete start-to-end []byte of all
+			// overlaps of the range to the callback (but in the next flush).
+			limit = match.from
+			break
+		}
+
+		if bufidx > match.from {
+			// This should never happen. It would mean that we wrote some of
+			// the buffer up to inside this match range. Maybe there's a bug
+			// in mergeOverlaps?
+			return fmt.Errorf("bufidx > match.from [%d > %d]", bufidx, match.from)
+		}
+
+		if bufidx < match.from {
+			// This is the non-matching part of the buffer before this match.
 			if _, err := r.dst.Write(r.buf[bufidx:match.from]); err != nil {
 				return err
 			}
-			fallthrough
+		}
 
-		case bufidx == match.from:
-			// A redacted range.
-			// Write a r.subst instead of the redacted range.
-			if _, err := r.dst.Write(r.subst); err != nil {
+		// Now handle the match itself.
+		// Call the replacement callback to get a replacement.
+		if repl := r.replacement(r.buf[match.from:match.to]); len(repl) > 0 {
```

Comment:
is "replace the thing matched with an empty string" (ie delete the matched string) a valid use case? i can see the semantics there being a little confusing if the user tried to do that

Reply:
That should be supported - `replacement` should return `nil`. The `if` on this line is just to skip calling `r.dst.Write` with nil or empty.

Reply:
sweet as